Correcting job_count usage

mghibaudi opened the pull request:
- The default value was overriding the local users settings (intentionally).
- The current priority is: command line -> application -> general
- Note: the environment variable in the `qstone.sh` is not referenced in the code

petergthatsme commented:
thanks @mghibaudi for taking a stab at this; and sorry for being slow. 
As far as i can tell there might still be an issue.
Although now `job_count` is respected, and even can be overwritten if what's passed in `user` is different than what's given in `environment`, if there are multiple users, and one of them overwrites `job_count`, then the same thing is set for the other user, even though they should be using the option given in `environment` instead. Here is a config that reproduces this for me:

```
{
  "environment": { 
     "project_name": "proj_name",
     "scheduling_mode" : "LOCK",
	 "job_count": 20,
     "qpu" : {
        "mode" : "EMULATED"
     },
     "connectivity": {
     	"mode": "RIGETTI",
		"qpu" : {
            "ip_address": "http://127.0.0.1",
            "port": 5000
		},
		"compiler": {
            "ip_address": "tcp://127.0.0.1",
            "port": 5555
		},
        "target": "10q-qvm"
     },
     "timeouts" : {
         "http": 2000,
         "lock": 5000
     } 
  },
  "jobs": [
    {
      "type": "VQE",
      "qubits": [4, 6],
      "num_shots": [100, 200],
      "walltime" : 10,
      "nthreads" : 4
    }
  ],
  "users": [
    {
      "user": "user0",
	  "job_count": 5,
      "computations": {
        "VQE": 1
      }
    },
    {
      "user": "user1",
      "computations": {
        "VQE": 1
      }
    }
  ]
}
```
So here `user0` should get 5 jobs, but `user1` should get 20.... 


Maybe also worth pointing out, as far as i remember we thought that it could make sense to essentially get rid of the `weight` in users (i don't include those in the config above), but then different apps for each user would get some weight that should add up to one (ie. 100%). Currently i think those may not get fully respected; also writing down something like: 
```
    {
      "user": "user1",
      "computations": {
        "VQE": 0.3
      }
    }
```
does not raise an error, which i think it should (?) - i.e. the total portions of each app for each user should equal to 1, whereas here it's just 0.3 
(or I may be misremembering what we discussed about how we want that to work? )

mghibaudi commented:
> thanks @mghibaudi for taking a stab at this; and sorry for being slow. As far as i can tell there might still be an issue. Although now `job_count` is respected, and even can be overwritten if what's passed in `user` is different than what's given in `environment`, if there are multiple users, and one of them overwrites `job_count`, then the same thing is set for the other user, even though they should be using the option given in `environment` instead. Here is a config that reproduces this for me:
> 
> ```
> {
>   "environment": { 
>      "project_name": "proj_name",
>      "scheduling_mode" : "LOCK",
> 	 "job_count": 20,
>      "qpu" : {
>         "mode" : "EMULATED"
>      },
>      "connectivity": {
>      	"mode": "RIGETTI",
> 		"qpu" : {
>             "ip_address": "http://127.0.0.1",
>             "port": 5000
> 		},
> 		"compiler": {
>             "ip_address": "tcp://127.0.0.1",
>             "port": 5555
> 		},
>         "target": "10q-qvm"
>      },
>      "timeouts" : {
>          "http": 2000,
>          "lock": 5000
>      } 
>   },
>   "jobs": [
>     {
>       "type": "VQE",
>       "qubits": [4, 6],
>       "num_shots": [100, 200],
>       "walltime" : 10,
>       "nthreads" : 4
>     }
>   ],
>   "users": [
>     {
>       "user": "user0",
> 	  "job_count": 5,
>       "computations": {
>         "VQE": 1
>       }
>     },
>     {
>       "user": "user1",
>       "computations": {
>         "VQE": 1
>       }
>     }
>   ]
> }
> ```
> 
> So here `user0` should get 5 jobs, but `user1` should get 20....
> 
> Maybe also worth pointing out, as far as i remember we thought that it could make sense to essentially get rid of the `weight` in users (i don't include those in the config above), but then different apps for each user would get some weight that should add up to one (ie. 100%). Currently i think those may not get fully respected; also writing down something like:
> 
> ```
>     {
>       "user": "user1",
>       "computations": {
>         "VQE": 0.3
>       }
>     }
> ```
> 
> does not raise an error, which i think it should (?) - i.e. the total portions of each app for each user should equal to 1, whereas here it's just 0.3 (or I may be misremembering what we discussed about how we want that to work? )

Good catch on both. I've addressed both issues (typo the first one, added a checker post json parsing for the latter)

mghibaudi commented:
@petergthatsme - addressed both points, merging the PR - happy to reopen if you still find issues on this. For the second issue I've managed to generate the test to cover it, the first one require quite a bit of redesign of the test environment to be effective/neat. I'm looking into this (it will be a separate PR)
